Fix: Compiler warnings about memsetting non-trivial classes

LordAro opened the pull request:
Noticed a couple of warnings with GCC 10:

```
/home/lordaro/dev/openttd/src/network/network_gui.cpp: In member function ‘virtual void NetworkLobbyWindow::OnClick(Point, int, int)’:
/home/lordaro/dev/openttd/src/network/network_gui.cpp:1583:61: warning: ‘void* memset(void*, int, size_t)’ clearing an object of type ‘struct NetworkCompanyInfo’ with no trivial copy-assignment; use assignment or value-initialization instead [-Wclass-memaccess]
 1583 |     memset(this->company_info, 0, sizeof(this->company_info));
      |                                                             ^
In file included from /home/lordaro/dev/openttd/src/network/network_gui.cpp:15:
/home/lordaro/dev/openttd/src/network/network_gui.h:28:8: note: ‘struct NetworkCompanyInfo’ declared here
   28 | struct NetworkCompanyInfo : NetworkCompanyStats {
      |        ^~~~~~~~~~~~~~~~~~
/home/lordaro/dev/openttd/src/cargotype.cpp: In function ‘void SetupCargoForClimate(LandscapeID)’:
/home/lordaro/dev/openttd/src/cargotype.cpp:45:54: warning: ‘void* memset(void*, int, size_t)’ clearing an object of type ‘struct CargoSpec’ with no trivial copy-assignment; use assignment or value-initialization instead [-Wclass-memaccess]
   45 |  memset(CargoSpec::array, 0, sizeof(CargoSpec::array));
      |                                                      ^
In file included from /home/lordaro/dev/openttd/src/cargotype.cpp:11:
/home/lordaro/dev/openttd/src/cargotype.h:55:8: note: ‘struct CargoSpec’ declared here
   55 | struct CargoSpec {
      |        ^~~~~~~~~
```

LordAro committed:
Fix: Compiler warnings about memsetting non-trivial classes

LordAro committed:
Codechange: Add WARN_FORMAT to vseprintf and fix the cascade of warnings that followed

michicc reviewed (dismissed):
Looks like the warning is because Money (aka OverflowSaveInt) has a default constructor.

LordAro committed:
Fix: Warning about using the wrong enum type

LordAro committed:
Codechange: Only apply FORTIFY_SOURCE in non-debug builds

It requires -O1 (or greater) and GCC spews out warnings if you try using it with -O0

LordAro committed:
Codechange: Make sure script enums are the same size as their normal counterparts

michicc reviewed (approved):
Not approving again 😈
